src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java

```diff
@@ -0,0 +1,118 @@
+	protected final Project project;
+	private final Collection<Configuration> annotationProcessorConfigurations;
+	protected final TaskCollection<T> invokerTasks;
+
+	protected AnnotationProcessorInvoker(Project project,
+										Collection<Configuration> annotationProcessorConfigurations,
+										TaskCollection<T> invokerTasks) {
+		this.project = project;
+		this.annotationProcessorConfigurations = annotationProcessorConfigurations;
+		this.invokerTasks = invokerTasks;
+	}
+
+	protected abstract void passArgument(T compileTask, String key, String value);
+
+	protected abstract File getDestinationDir(T task);
+
+	protected final String getRefmapDestination(T task, LoomGradleExtension extension) throws IOException {
+		return new File(getDestinationDir(task), extension.getRefmapName()).getCanonicalPath();
+	}
+
+	private void passMixinArguments(T task) {
+		try {
+			LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
+			Map<String, String> args = new HashMap<String, String>() {{
+					put("inMapFileNamedIntermediary", extension.getMappingsProvider().tinyMappings.getCanonicalPath());
+					put("outMapFileNamedIntermediary", extension.getMappingsProvider().mappingsMixinExport.getCanonicalPath());
+					put("outRefMapFile", getRefmapDestination(task, extension));
+					put("defaultObfuscationEnv", "named:intermediary");
+				}};
+
+			project.getLogger().info("Outputting refmap to dir: " + getDestinationDir(task) + " for compile task: " + task);
+			args.forEach((k, v) -> passArgument(task, k, v));
+		} catch (IOException e) {
+			project.getLogger().error("Could not configure mixin annotation processors", e);
+		}
+	}
+
+	public void configureMixin() {
+		ConfigurationContainer configs = project.getConfigurations();
+		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
+
+		if (!extension.ideSync()) {
+			for (Configuration processorConfig : annotationProcessorConfigurations) {
+				project.getLogger().info("Adding mixin to classpath of AP config: " + processorConfig.getName());
+				// Pass named MC classpath to mixin AP classpath
+				processorConfig.extendsFrom(
+								configs.getByName(Constants.MINECRAFT_NAMED),
+								configs.getByName(Constants.MOD_COMPILE_CLASSPATH_MAPPED),
+								configs.getByName(Constants.MAPPINGS_FINAL)
+				);
+
+				// Add Mixin and mixin extensions (fabric-mixin-compile-extensions pulls mixin itself too)
+				project.getDependencies().add(processorConfig.getName(),
+								"net.fabricmc:fabric-mixin-compile-extensions:" + Constants.MIXIN_COMPILE_EXTENSIONS_VERSION);
+			}
+		}
+
+		for (T task : invokerTasks) {
+			passMixinArguments(task);
+		}
+	}
+
+	static Stream<SourceSet> getNonTestSourceSets(Project project) {
+		return project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets()
+						.stream()
+						.filter(sourceSet -> !sourceSet.getName().equals("test"));
```

Comment:
Why is this needed? Should it be excluding custom test source sets like `testMod` in fabric api?

Reply:
There is similar code in the old code that excludes sourcesets that include the work `test` in them, so I thought I would do the same. I made it only target specifically `test` though because having mixins arbitraily not work because your sourceset has the word `test` in them (like `testMod`) doesn't make sense.

Reply:
Ok, that sounds resonable. As they are needed on the testMod sourceset so that makes sense. Does it cause any harm if they are ran on the test source set?

Reply:
No, it doesn't cause any issues. This code was just here to make it more similar to the intention of the old behavior.